lib/ipam/ipam.go

```diff
@@ -195,9 +195,52 @@ func (c ipamClient) getBlockFromAffinity(ctx context.Context, aff *model.KVPair)
 	return b, nil
 }
 
-func (c ipamClient) autoAssign(ctx context.Context, num int, handleID *string, attrs map[string]string, pools []net.IPNet, version ipVersion, host string) ([]net.IP, error) {
+func (c ipamClient) autoAssign(ctx context.Context, num int, handleID *string, attrs map[string]string, requestedPools []net.IPNet, version ipVersion, host string) ([]net.IP, error) {
+	// Start by sanitizing the requestedPools.
+	pools := []net.IPNet{}
 
-	// Start by trying to assign from one of the host-affine blocks.  We
+	// Get all the configured pools.
```

Comment:
Probably worth encapsulating the pool determination logic into its own function for easier testing and to clean up the `autoAssign` code a bit.

Reply:
This code then becomes:

```go
pools, err := c.determinePools(requestedPools)
if err != nil { ... }
```